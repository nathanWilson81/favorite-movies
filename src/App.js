import React, { Component } from 'react'
import styled from 'styled-components'
import { map, equals, reject, filter, find } from 'lodash/fp'
import { MovieCards, AdditionalInfoModal, Dropdown } from './components'
import { getGenres, getNewMovie } from './constants'
import { persistToLocalStorage, getFromLocalStorage } from './localStorage'

const Wrapper = styled.div`
  height: 100vh;
  display: flex;
  align-items: center;
  flex-direction: column;
  background: #f5f5f5;
`

const Header = styled.div`
  min-height: 140px;
  width: 60%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background: #fff;
  padding: 0.5rem;
  margin-bottom: 1rem;
  margin-top: 1rem;
  box-shadow: 3px 3px 3px rgba(0, 0, 0, 0.03);
`

const CardContainer = styled.div`
  width: 60%;
  overflow-x: auto;
`

const SortContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`

const SortText = styled.h3`
  margin-bottom: 0;
  margin-top: 0;
  margin-right: 1rem;
`

const Button = styled.button`
  border-radius: 2rem;
  height: 2rem;
  margin-top: 1rem;
`

class App extends Component {
  state = {
    movies: [],
    genreSort: 'All',
    selectedMovie: null,
    additionalInfoModalOpen: false
  }

  componentDidMount() {
    const localStorageMovies = getFromLocalStorage('movies')
    if (localStorageMovies) {
      this.setState({
        movies: localStorageMovies
      })
    }
  }

  addNewMovie = genre =>
    this.setState({
      movies: this.state.movies.concat(getNewMovie(genre))
    })

  deleteMovie = movie =>
    this.setState(
      {
        movies: reject(equals(movie), this.state.movies)
      },
      () => persistToLocalStorage('movies', this.state.movies)
    )

  updateMovie = (movie, update) =>
    this.setState(
      {
        movies: map(
          val => (equals(val, movie) ? { ...movie, ...update } : val),
          this.state.movies
        )
      },
      () => persistToLocalStorage('movies', this.state.movies)
    )

  openAdditionalInfoModal = movie =>
    this.setState({
      selectedMovie: movie,
      additionalInfoModalOpen: true
    })

  closeAdditionalInfoModal = () =>
    this.setState({ selectedMovie: null, additionalInfoModalOpen: false })

  render() {
    const visibleMovies =
      this.state.genreSort === 'All'
        ? this.state.movies
        : filter(
            movie => movie.movieGenre === this.state.genreSort,
            this.state.movies
          )
    return (
      <React.Fragment>
        <AdditionalInfoModal
          isOpen={this.state.additionalInfoModalOpen}
          movie={
            this.state.selectedMovie &&
            find(
              movie => this.state.selectedMovie.id === movie.id,
              this.state.movies
            )
          }
          onClose={this.closeAdditionalInfoModal}
          updateMovie={this.updateMovie}
        />
        <Wrapper>
          <Header>
            <h1>My Favorite Movies</h1>
            <SortContainer>
              <SortText>Sort By Genre: </SortText>
              <Dropdown
                value={this.state.genreSort}
                values={[]
                  .concat({ label: 'All', value: 'All' })
                  .concat(getGenres())}
                onChange={e => {
                  this.setState({ genreSort: e.target.value })
                }}
              />
            </SortContainer>
            <Button onClick={() => this.addNewMovie(this.state.genreSort)}>
              Add New Movie
            </Button>
          </Header>
          <CardContainer>
            <MovieCards
              movies={visibleMovies}
              deleteMovie={this.deleteMovie}
              updateMovie={this.updateMovie}
              openAdditionalInfoModal={this.openAdditionalInfoModal}
              starCount={5}
              genres={getGenres()}
            />
          </CardContainer>
        </Wrapper>
      </React.Fragment>
    )
  }
}

export default App
