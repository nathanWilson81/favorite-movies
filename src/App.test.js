import React from 'react'
import renderer from 'react-test-renderer'
import App from './App'

describe('App', () => {
  let component
  beforeEach(() => {
    component = renderer.create(<App />)
  })

  it('should match a snapshot', () => {
    const snapshot = component.toJSON()
    expect(snapshot).toMatchSnapshot()
  })
})
