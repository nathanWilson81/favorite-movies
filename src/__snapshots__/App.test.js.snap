// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match a snapshot 1`] = `
<div
  className="sc-dnqmqq kLNolu"
>
  <div
    className="sc-iwsKbI bOmJaN"
  >
    <h1>
      My Favorite Movies
    </h1>
    <div
      className="sc-gqjmRU esGRdK"
    >
      <h3
        className="sc-VigVT htzFyS"
      >
        Sort By Genre: 
      </h3>
      <select
        className="sc-bdVaJa fqmIub"
        onChange={[Function]}
        value="All"
      >
        <option
          value="All"
        >
          All
        </option>
        <option
          value="Action"
        >
          Action
        </option>
        <option
          value="Adventure"
        >
          Adventure
        </option>
        <option
          value="Animation"
        >
          Animation
        </option>
        <option
          value="Childrens"
        >
          Childrens
        </option>
        <option
          value="Comedy"
        >
          Comedy
        </option>
        <option
          value="Crime"
        >
          Crime
        </option>
        <option
          value="Documentary"
        >
          Documentary
        </option>
        <option
          value="Drama"
        >
          Drama
        </option>
        <option
          value="Horror"
        >
          Horror
        </option>
        <option
          value="Family"
        >
          Family
        </option>
        <option
          value="Fantasy"
        >
          Fantasy
        </option>
        <option
          value="Historical"
        >
          Historical
        </option>
        <option
          value="Medical"
        >
          Medical
        </option>
        <option
          value="Musical"
        >
          Musical
        </option>
        <option
          value="Paranormal"
        >
          Paranormal
        </option>
        <option
          value="Romance"
        >
          Romance
        </option>
        <option
          value="Sport"
        >
          Sport
        </option>
        <option
          value="Science fiction"
        >
          Science fiction
        </option>
        <option
          value="Thriller"
        >
          Thriller
        </option>
        <option
          value="War"
        >
          War
        </option>
        <option
          value="Western"
        >
          Western
        </option>
      </select>
    </div>
    <button
      className="sc-jTzLTM eoEWZr"
      onClick={[Function]}
    >
      Add New Movie
    </button>
  </div>
  <div
    className="sc-gZMcBi czVSTk"
  />
</div>
`;
